File: 01.Python/0918.py
from typing import List

class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        localMax = nums[0]
        globalMax = nums[0]

        localMin = nums[0]
        globalMin = nums[0]

        totalSum = nums[0]

        for i in range(1, len(nums)):
            localMax = max(nums[i], localMax + nums[i])
            globalMax = max(globalMax, localMax)

            localMin = min(nums[i], nums[i] + localMin)
            globalMin = min(globalMin, localMin)

            totalSum += nums[i]

        if globalMin == totalSum:
            return globalMax

        wrappingSum = totalSum - globalMin
        result = max(globalMax, wrappingSum)

        return result

sol = Solution()

nums1 = [1, -2, 3, -2]
resutlt1 = sol.maxSubarraySumCircular(nums1)
print(resutlt1)

nums2 = [5, -3, 5]
result2 = sol.maxSubarraySumCircular(nums2)
print(result2)

nums3 = [-3, -2, -3]
result3 = sol.maxSubarraySumCircular(nums3)
print(result3)

File: 03.CPP/0918.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int maxSubarraySumCircular(std::vector<int>& nums) {
        int localMax = nums[0];
        int globalMax = nums[0];

        int localMin = nums[0];
        int globalMin = nums[0];

        int totalSum = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            localMax = std::max(nums[i], nums[i] + localMax);
            globalMax = std::max(localMax, globalMax);

            localMin = std::min(nums[i], nums[i] + localMin);
            globalMin = std::min(localMin, globalMin);

            totalSum += nums[i];
        }

        if (totalSum == globalMin) {
            return globalMax;
        }

        int wrappingSum = totalSum - globalMin;
        int result = std::max(globalMax, wrappingSum);

        return result;
    }
};